cloud.ganttproject.colloboque/src/main/kotlin/cloud/ganttproject/colloboque/ColloboqueServer.kt

```diff
@@ -36,39 +36,96 @@ import java.util.*
 import javax.sql.DataSource
 import net.sourceforge.ganttproject.storage.InitRecord
 import net.sourceforge.ganttproject.storage.InputXlog
+import net.sourceforge.ganttproject.storage.XlogRecord
+import java.time.LocalDateTime
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.write
+
+private typealias ProjectRefid = String
+
+class ColloboqueServerException: Exception {
+  constructor(message: String): super(message)
+  constructor(message: String, cause: Throwable?): super(message, cause)
+}
 
 class ColloboqueServer(
   private val dataSourceFactory: (projectRefid: String) -> DataSource,
   private val initInputChannel: Channel<InitRecord>,
   private val updateInputChannel: Channel<InputXlog>) {
+  private val refidToDataSource: MutableMap<ProjectRefid, DataSource> = mutableMapOf()
```

Comment:
I suggest keeping this project to data source mapping in the factory, to make code in this file easier.